File: backend/core/ratelimit.py
"""Tiny in-process rate limiter (fixed-interval cooldown).

Used to throttle expensive Meta sync operations per tenant so an impatient
user (or a runaway client) cannot hammer the Graph API. In-process only
(per worker) -- adequate as an accidental-abuse guard; for strict global
limits across workers back this with Redis later.
"""
from __future__ import annotations

import os
import threading
import time
from typing import Dict, Tuple

_lock = threading.Lock()
_last_call: Dict[str, float] = {}


def check_cooldown(key: str, cooldown_seconds: float) -> Tuple[bool, int]:
    """Return ``(allowed, retry_after_seconds)``.

    Allows the call if at least ``cooldown_seconds`` have elapsed since the
    last *allowed* call for ``key``. Records the timestamp only when allowed,
    so a blocked caller does not extend its own cooldown. ``cooldown_seconds
    <= 0`` disables the limit.
    """
    if cooldown_seconds <= 0:
        return True, 0
    now = time.monotonic()
    with _lock:
        last = _last_call.get(key)
        if last is not None:
            elapsed = now - last
            if elapsed < cooldown_seconds:
                return False, max(1, int(round(cooldown_seconds - elapsed)))
        _last_call[key] = now
    return True, 0


def sync_cooldown_seconds() -> float:
    """Cooldown between campaign syncs per tenant (env ``SYNC_RATE_LIMIT_SECONDS``)."""
    try:
        return float(os.environ.get("SYNC_RATE_LIMIT_SECONDS", "15"))
    except (TypeError, ValueError):
        return 15.0


def reset() -> None:  # for tests
    with _lock:
        _last_call.clear()

File: apps/web/lib/rate-limit.ts
/**
 * In-process fixed-interval cooldown limiter (per worker/instance). Used to
 * throttle expensive Meta syncs per tenant. For strict global limits across
 * serverless instances, back this with Upstash/Redis later.
 */
const lastCall = new Map<string, number>();

export interface CooldownResult {
  allowed: boolean;
  retryAfter: number; // seconds
}

export function checkCooldown(key: string, cooldownSeconds: number): CooldownResult {
  if (cooldownSeconds <= 0) return { allowed: true, retryAfter: 0 };
  const now = Date.now();
  const prev = lastCall.get(key);
  if (prev !== undefined) {
    const elapsed = (now - prev) / 1000;
    if (elapsed < cooldownSeconds) {
      return { allowed: false, retryAfter: Math.max(1, Math.ceil(cooldownSeconds - elapsed)) };
    }
  }
  lastCall.set(key, now);
  return { allowed: true, retryAfter: 0 };
}

export function syncCooldownSeconds(): number {
  const raw = process.env.SYNC_RATE_LIMIT_SECONDS;
  const v = raw === undefined ? 15 : Number(raw);
  return Number.isFinite(v) ? v : 15;
}

export function resetRateLimit(): void {
  lastCall.clear();
}
